### PR TITLE
feat: Add TowerStatsFeature and TowerStatsOverlay

### PR DESCRIPTION
**Pull request by kristofbolyai**
## Overlay
image

## Chat log
image

### CODE REVIEW

File: common/src/main/java/com/wynntils/functions/WarFunctions.java [hunk 1]
```diff
@@ -22,4 +26,126 @@ public Double getValue(FunctionArguments arguments) {
...
+            if (initialTowerStateOpt.isEmpty()) return -1;
+
+            return initialTowerStateOpt.get().defense();
+        }
+    }
+
+    public static class InitialTowerDamageFunction extends Function<CappedValue> {
+        @Override
+        public CappedValue getValue(FunctionArguments arguments) {
+            // Technically this is a RangedValue, but functions don't support RangedValue
```
Comment (magicus) [inline common/src/main/java/com/wynntils/functions/WarFunctions.java]: Should we fix that?

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/functions/WarFunctions.java]: This should be fine for the time being imo.

Comment (magicus) [reply common/src/main/java/com/wynntils/functions/WarFunctions.java]: Ok.

File: common/src/main/java/com/wynntils/functions/generic/RangedFunctions.java
Comment (magicus) [inline common/src/main/java/com/wynntils/functions/generic/RangedFunctions.java:1]: Thanks. I think this is better to actually expose our "reality" in functions.

File: common/src/main/java/com/wynntils/functions/generic/StringFunctions.java [hunk 2]
```diff
@@ -43,6 +44,33 @@ public FunctionArguments.RequiredArgumentBuilder getRequiredArgumentsBuilder() {
         }
     }
 
+    public static class FormatRangedFunction extends GenericFunction<String> {
+        @Override
+        public String getValue(FunctionArguments arguments) {
+            RangedValue value = arguments.getArgument("value").getRangedValue();
+            return StringUtils.integerToShortString(value.low()) + "/" + StringUtils.integerToShortString(value.high());
```
Comment (magicus) [inline common/src/main/java/com/wynntils/functions/generic/StringFunctions.java]: We represent RangedValues as `<low>-<high>` in our toString method. I think we should do the same here. That is kind of the essential difference between a ranged value ("between low and high") and a capped value `<cur>/<max>` ("the value cur out of max").

**Comment by magicus** [comment]
Some of the more general changes (like in `StringUtil.java`, and the function value changes) perhaps should be extracted into a separate PR that targets main directly?

**Comment by kristofbolyai** [comment]
> Some of the more general changes (like in `StringUtil.java`, and the function value changes) perhaps should be extracted into a separate PR that targets main directly?

I think it's not necessary in this case, since main wouldn't be using any of the function changes. As for the StringUtil class, I think there is no rename there, just a int -> long type change, which should work implicitly. Also, in theory, main won't get anything "new", except for the item filter UI, until this PR is finished (which I am pushing hard for, as you can see, so it can be done before I  have all my exams).

File: common/src/main/java/com/wynntils/functions/generic/StringFunctions.java (hunk 2 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/functions/generic/StringFunctions.java]:
```suggestion
            return StringUtils.integerToShortString(value.low()) + "-" + StringUtils.integerToShortString(value.high());
```

Indeed! I just copied the method, and forgot to change it. Thanks!

**Review by magicus — APPROVED** [review]

**Pull request merged**